Fuseki 5 & Nextgraph Integration

SlyRock opened the pull request:
closes :
#1250 Upgrade Fuseki to v5
#1268 Refactor DatasetService
#1263 Refactor TripleStoreService to be compatible with NextGraph

- Extended docker image for fuseki5 to handle permissions issue, (see [jena/docker/fuseki5-docker](https://github.com/assemblee-virtuelle/semapps/pull/1417/files#diff-8fc25bd1fae1ec88f0c2dd46575119fec7a6fca999815deae982d571da8096e0))
- Full test suite for triplestore service and subservice with adequate coverage to ensure fuseki v5 and nextgraph compatibility (see src/middleware/tests/triplestore/triplestore.test.ts)
- Refactor of triplestore service with an implementation of an adapter pattern to allow application to use either fuseki5 or nextgraph with a unique interface (see src/middleware/packages/triplestore)
- Refactor of the dataset subservice to work with the adapter pattern (see src/middleware/packages/triplestore/subservices/dataset.ts) 
- Added named-graph service, also using the adapter pattern (see src/middleware/packages/triplestore/subservices/named-graph.ts)
- Adapt ldp service tests to use NextGraph

SlyRock committed:
Full test suite for triplestore service and subservice with adequate coverage
Refactor triplestore service to support multiple backends (Fuseki and NextGraph)
TODO before review : Environmentalize the nextgraph settings.

SlyRock committed:
Remove yarn.lock and package-lock.json files, clean up middleware tests by deleting unused files, and update package.json dependencies for triplestore and tests. Rename triplestore package for consistency.
Migrate tests to ts.

srosset81 commented on src/middleware/packages/triplestore/service.ts:
```diff
@@ -19,28 +13,30 @@ const { MoleculerError } = Errors;
-    url: null,
-    user: null,
-    password: null,
     mainDataset: null,
```
Rename to `defaultDataset`

srosset81 commented on src/middleware/tests/triplestore/initialize.ts:
```diff
@@ -0,0 +1,70 @@
+        'did:ng:o:5ZwPgEib6okmEbVlWRJVfGUNnbdtmQpC_x1uTy9wjcoA:v:asrmmGCr1WTq3oAGkgtVwUxsJgA5MIsV2FIYhDRyPagA',
+      serverPeerId: 'zT_iEzpHeO5znVU9ZYcvenJjb8pWrRWFzEO6eUE_SrAA',
+      adminUserKey: 'dwtQ9wWEovJwv6_3VArHKHRyr_zLAuR2_bFB1LiLfqEA',
+      clientPeerKey: 'ryv9v1Y3jJqdQYH-_rMxGTGyDtC_eOaA0a4ibRLhmX4A',
```
To put in env variables

srosset81 commented on src/middleware/tests/triplestore/initialize.ts:
```diff
@@ -0,0 +1,70 @@
+      url: CONFIG.SPARQL_ENDPOINT,
+      user: CONFIG.JENA_USER,
+      password: CONFIG.JENA_PASSWORD,
+      mainDataset: CONFIG.MAIN_DATASET
```
`mainDataset` is not a parameter of the FusekiAdapter.

srosset81 commented on src/middleware/packages/triplestore/adapters/nextgraph.ts:
```diff
@@ -0,0 +1,301 @@
+
+  constructor(settings: NextGraphAdapterSettings) {
+    super();
+    if (!settings.serverPeerId) throw new Error('Server peer id is required');
+    if (!settings.adminUserKey) throw new Error('Admin user key is required');
+    if (!settings.clientPeerKey) throw new Error('Client peer key is required');
+    if (!settings.serverAddr) throw new Error('Server address is required');
```
Shouldn't this be saved in `this.nextgraphConfig` directly ?

srosset81 commented on src/middleware/packages/triplestore/subservices/dataset.ts:
```diff
@@ -1,141 +1,63 @@
-    user: null,
-    password: null,
-    fusekiBase: null
+    backend: null as AdapterInterface | null
```
`adapter`, not `backend`

SlyRock committed:
Refactor triplestore actions to remove 'webId' parameter from dropAll, query, and update schemas. Add AdapterInterface import in service.ts for improved type handling.

srosset81 commented on src/middleware/packages/triplestore/package.json:
```diff
@@ -9,6 +9,7 @@
     "@semapps/mime-types": "1.1.4",
     "jsonld": "^3.3.2",
     "negotiator": "^0.6.2",
+    "nextgraph": "file:../../../../../../../NG/nextgraph-rs/ng-sdk-js/pkg-node",
```
Use real nextgraph package ?

SlyRock replied:
Indeed, this will be changed ultimately before merging the PR.
It will have to be changed locally to run the nextgraph adapter, until we have a proper sync between the released SDK and dockerized NG server.

srosset81 commented on src/middleware/packages/triplestore/README.md:
```diff
@@ -1,5 +1,81 @@
+    password: 'password',
+    mainDataset: 'main'
+  }
+}
```
The whole README.me is not up-to-date

srosset81 commented on src/middleware/packages/triplestore/README.md:
```diff
@@ -1,5 +1,81 @@
+## Dependencies
+
+- **Fuseki**: Uses `node-fetch` for HTTP communication
+- **NextGraph**: Uses the `nextgraph` SDK (optional dependency)
```
The dependency is not optional, otherwise it needs to be an external dependency ;) (but we can do without that)

srosset81 commented on src/middleware/tests/triplestore/triplestore.test.ts:
```diff
@@ -0,0 +1,549 @@
+      });
+    } catch (error) {
+      // Dataset might not exist, ignore error
+    }
```
In all other test suites, we put this in `beforeAll` so that we can look at what was done after the test suite is run.
Also, if the test is interrupted, it will not cause problems when restarting the tests

srosset81 commented on src/middleware/tests/triplestore/triplestore.test.ts:
```diff
@@ -0,0 +1,549 @@
+        });
+      } catch (error) {
+        // Intentionally ignore errors if dataset does not exist
+      }
```
If dataset deletion is in done in `beforeAll`, this is not necessary.

srosset81 commented on src/middleware/tests/triplestore/triplestore.test.ts:
```diff
@@ -0,0 +1,549 @@
+    test('SELECT query with JSON result', async () => {
+      const result = await broker.call('triplestore.query', {
+        query: 'SELECT ?s ?p ?o WHERE { ?s ?p ?o } LIMIT 5',
+        // accept: 'application/json',
```
All the commented `accept` params can be deleted

srosset81 commented on src/middleware/tests/triplestore/triplestore.test.ts:
```diff
@@ -0,0 +1,549 @@
+        // accept: 'application/json',
+        dataset: testDataset
+      });
+      expect(typeof result).toBe('object');
```
IMO we should check the actual answer

srosset81 commented on src/middleware/tests/triplestore/triplestore.test.ts:
```diff
@@ -0,0 +1,549 @@
+        query: sparqlObject,
+        dataset: testDataset
+      });
+      expect(Array.isArray(result)).toBeTruthy();
```
Here also we should check the actual result

srosset81 commented:
Please update also the documentation of the triplestore package in the website

SlyRock committed:
Refactor triplestore service to replace 'mainDataset' with 'defaultDataset' for consistency across actions and settings. Update related actions to utilize the new default dataset configuration.
Add tests for named-graph subservice : Highlights issues with the service itself, that therefore needs fixing

SlyRock committed:
Force explicit creation of named graph via declarative triple insertion

srosset81 committed:
New BinaryService using adapters

srosset81 committed:
Remove the ldp.resource.upload action

srosset81 committed:
Cleanup

srosset81 committed:
Add NgBinaryAdapter

srosset81 committed:
Add migration tool

srosset81 committed:
Add documentation

srosset81 committed:
Add link in docs

srosset81 committed:
Merge branch '2.0' into binary-adapter

SlyRock committed:
Enhance named graph functionality in triplestore service by implementing create, check existence, clear, and delete operations in adapters. Update NamedGraphService to utilize these new adapter methods, ensuring compatibility with both Fuseki and NextGraph. Add tests to validate named graph operations across different backends.

SlyRock committed:
Update NextGraph adapter documentation to clarify client peer key generation process. Modify triplestore tests to ensure proper dataset cleanup after tests.

srosset81 committed:
Fix tests, remove iKnowWhatImDoing param

srosset81 committed:
Merge branch '2.0' into feature/nextgraph-integration-ts

srosset81 committed:
Tests passing except for NextGraph not accepting HTTP URIs as graph name

srosset81 committed:
Don't use a baseUrl for named graphs

srosset81 committed:
Clean up and fix tests

srosset81 committed:
Working account test

srosset81 committed:
Cleanup session handling

srosset81 committed:
Most LDP tests working except controlled containers/resources and API

srosset81 committed:
Remove console.log

srosset81 committed:
All LDP tests running except binary and blank nodes

srosset81 committed:
Merge branch 'feature/nextgraph-integration-ts' into binary-adapter

srosset81 committed:
Fix all LDP tests

srosset81 committed:
Merge pull request #1428 from assemblee-virtuelle/binary-adapter

New LdpBinaryService using adapters

SlyRock committed:
Update Docker Compose configuration for ng_tests service and enhance triplestore initialization with environmentalized settings. Add new dependencies in package.json for improved test functionality.

SlyRock committed:
Update .env and docker-compose configurations for ng_tests service to include new Nextgraph initialization parameters and adjust port mappings for improved functionality.

SlyRock committed:
Merge branch 'feature/nextgraph-integration-ts' of github.com:assemblee-virtuelle/semapps into feature/nextgraph-integration-ts

SlyRock committed:
Refactor .env and docker-compose configurations for ng_tests service t

srosset81 committed:
Use NextGraph env variables correctly

srosset81 committed:
Fix tests

SlyRock committed:
Docker image extension for fuseki 5, in order to fix the permissions issue

SlyRock committed:
Merge branch 'feature/nextgraph-integration-ts' of git@github.com:assemblee-virtuelle/semapps.git

srosset81 committed:
Update nextgraph docker config

srosset81 committed:
Fixes

srosset81 committed:
Use new NG package and add Makefile
